File: runme/mcgen/local/datan/datsrc/auxdrg.c
#include <stdlib.h>
#include "datsrc.h"

#if defined(MAKELIB)
extern struct {
    double dx[10];
    double w[30]; /* was [10][3] */
    logical lev[10];
} cmauxc;
#endif

void LIBFUNC auxdrg_(double P_T *x, double P_T *eta, integer const P_T *mm,
             integer const P_T *n, integer const P_T *nr,
             integer const P_T *nred, integer const P_T *list,
             double P_T *e, logical P_T *ok)
{
#if !defined(MAKELIB)
    /* Initialized data */

    static double dx[10] = { .0256,.0192,.0128,.0096,.0064,.0048,.0032,
            .0024,.0016,.0012 };
    static logical lev[10] = { TRUE_,FALSE_,TRUE_,FALSE_,TRUE_,FALSE_,TRUE_,
            FALSE_,TRUE_,FALSE_ };
    static double w[30] /* was [10][3] */ = { 0.0,1.3333333333333333,
            .33333333333333333,1.0666666666666667,.066666666666666667,
            1.0158730158730159,.015873015873015873,1.003921568627451,
            .0039215686274509804,0.0,2.2857142857142857,1.2857142857142857,
            1.1636363636363636,.16363636363636364,1.0364372469635628,
            .036437246963562753,1.0088669950738916,.0088669950738916256,
            1.0022021042329337,.0022021042329336922,1.8,.8,1.125,.125,
            1.0285714285714286,.028571428571428571,1.006993006993007,
            .006993006993006993,1.0017391304347826,.0017391304347826087 };
#else
    double *w = cmauxc.w, *dx = cmauxc.dx;
    logical *lev = cmauxc.lev;
#endif

    /* System generated locals */
    integer e_dim1, e_offset, i__1, i__2, i__3;
    double d__1;

    /* Local variables */
    integer i, k, l, m, i2, il, im, iy;
    double fminus, fplus, del, xsav, h;
    logical lmt, lx;
    /* double a[10], t[100] (* was [10][10] *);*/
    double *a, *t;

    a = (double *)MEMALLOC(sizeof(double)*(100+10));
    if (!a) {
        MEMERROR("auxdrg_");
        *ok = FALSE_;
        return;
    }
    t = a + 10;

/* Computes the derivative f'(x) of f(x) at x = X. Based on */
/* H. Rutishauser, Ausdehnung des Rombergschen Prinzips */
/* (Extension of Romberg's Principle), Numer. Math. 5 (1963) 48-54 */
    /* Parameter adjustments */
    --eta;
    --list;
    --x;
    e_dim1 = *mm;
    e_offset = e_dim1 + 1;
    e -= e_offset;

    /* Function Body */
    *ok = TRUE_;
    l = *nr + *n;
    i__1 = *mm;
    for (im = 1; im <= i__1; ++im) {
        i2 = 0;
        i__2 = l;
        for (il = 1; il <= i__2; ++il) {
            del = 10.;
            if (il <= *nr) {
                if (list[il] != 0) {
                    ++i2;
                    lx = TRUE_;
                    xsav = x[il];
                } else {
                    goto L80;
                }
            } else {
                ++i2;
                lx = FALSE_;
                iy = il - *nr;
                xsav = eta[iy];
            }
            for (i = 1; i <= 10; ++i) {
                del *= .1;
                if (i == 10 || (d__1 = xsav + del * dx[9] - xsav, abs(d__1)) <
                         5e-8) {
                    *ok = FALSE_;
                    goto L99;
                }
                for (k = 0; k <= 9; ++k) {
                    h = del * dx[k];
                    if (lx) {
                        x[il] = xsav + h;
                    } else {
                        eta[iy] = xsav + h;
                    }
                    fplus = lsqgfn_(&eta[1], &x[1], n, nr, &im);
                    if (lx) {
                        x[il] = xsav - h;
                    } else {
                        eta[iy] = xsav - h;
                    }
                    fminus = lsqgfn_(&eta[1], &x[1], n, nr, &im);
                    if (lx) {
                        x[il] = xsav;
                    } else {
                        eta[iy] = xsav;
                    }
                    t[k] = (fplus - fminus) / (h + h);
                    a[k] = t[k];
                }
                if (a[0] >= a[9]) {
                    for (k = 0; k <= 9; ++k) {
                        a[k] = -a[k];
                    }
                }
                lmt = TRUE_;
                for (k = 1; k <= 9; ++k) {
                    h = a[k - 1] - a[k];
                    lmt = lmt && (h <= 1e-10 || abs(h) <=
                                  (d__1 = a[k], abs(d__1)) * 5e-8 + 1e-10);
                }
                if (lmt) {
                    goto L50;
                }
            }
L50:
            for (m = 1; m <= 9; ++m) {
                i__3 = 9 - m;
                for (k = 0; k <= i__3; ++k) {
                    if (lev[m]) {
                        t[k + m * 10] = w[m - 1] * t[k + 1 + (m - 1) * 10] -
                                w[m] * t[k + (m - 1) * 10];
                    } else if (lev[k]) {
                        t[k + m * 10] = w[m + 9] * t[k + 1 + (m - 1) * 10] -
                                w[m + 10] * t[k + (m - 1) * 10];
                    } else {
                        t[k + m * 10] = w[m + 19] * t[k + 1 + (m - 1) * 10] -
                                w[m + 20] * t[k + (m - 1) * 10];
                    }
                }
            }
            e[im + i2 * e_dim1] = t[90];
L80:
            ;
        }
    }
L99:
    MEMFREE((void *)a);
    return;
} /* auxdrg_ */

File: runme/mcgen/local/datan/datsrc/auxdrg.f
      SUBROUTINE AUXDRG(X,ETA,MM,N,NR,NRED,LIST,E,OK)
C Computes the derivative f'(x) of f(x) at x = X. Based on
C H. Rutishauser, Ausdehnung des Rombergschen Prinzips
C (Extension of Romberg's Principle), Numer. Math. 5 (1963) 48-54
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER (EPS = 5.D-8, EPSI=1.D-10, DELTA=10.D0, S=1.D-1)
      DIMENSION X(NR),ETA(N),E(MM,NRED+N),LIST(NR)
      DIMENSION DX(0:9),W(0:9,3),T(0:9,0:9),A(0:9)
      LOGICAL LEV(0:9),LMT
      LOGICAL OK,LX
      DOUBLE PRECISION LSQGFN
      DATA DX /0.0256D0, 0.0192D0, 0.0128D0, 0.0096D0, 0.0064D0,
     +0.0048D0, 0.0032D0, 0.0024D0, 0.0016D0, 0.0012D0/
      DATA (LEV(K),K=0,8,2) /5*.TRUE./
      DATA (LEV(K),K=1,9,2) /5*.FALSE./
      DATA W(1,1) /1.33333 33333 333333D+00/
      DATA W(3,1) /1.06666 66666 666667D+00/
      DATA W(5,1) /1.01587 30158 730159D+00/
      DATA W(7,1) /1.00392 15686 274510D+00/
      DATA W(2,1) /3.33333 33333 333333D-01/
      DATA W(4,1) /6.66666 66666 666667D-02/
      DATA W(6,1) /1.58730 15873 015873D-02/
      DATA W(8,1) /3.92156 86274 509804D-03/
      DATA W(0,2) /2.28571 42857 142857D+00/
      DATA W(2,2) /1.16363 63636 363636D+00/
      DATA W(4,2) /1.03643 72469 635628D+00/
      DATA W(6,2) /1.00886 69950 738916D+00/
      DATA W(8,2) /1.00220 21042 329337D+00/
      DATA W(1,2) /1.28571 42857 142857D+00/
      DATA W(3,2) /1.63636 36363 636364D-01/
      DATA W(5,2) /3.64372 46963 562753D-02/
      DATA W(7,2) /8.86699 50738 916256D-03/
      DATA W(9,2) /2.20210 42329 336922D-03/
      DATA W(0,3) /1.80000 00000 000000D+00/
      DATA W(2,3) /1.12500 00000 000000D+00/
      DATA W(4,3) /1.02857 14285 714286D+00/
      DATA W(6,3) /1.00699 30069 930070D+00/
      DATA W(8,3) /1.00173 91304 347826D+00/
      DATA W(1,3) /8.00000 00000 000000D-01/
      DATA W(3,3) /1.25000 00000 000000D-01/
      DATA W(5,3) /2.85714 28571 428571D-02/
      DATA W(7,3) /6.99300 69930 069930D-03/
      DATA W(9,3) /1.73913 04347 826087D-03/
      OK=.TRUE.
      L=NR+N
      DO 90 IM=1,MM
        I2=0
        DO 80 IL=1,L
          DEL=DELTA
          IF(IL.LE.NR) THEN
            IF(LIST(IL).NE.0) THEN
              I2=I2+1
              LX=.TRUE.
              XSAV=X(IL)
            ELSE
              GO TO 80
            END IF
          ELSE
            I2=I2+1
            LX=.FALSE.
            IY=IL-NR
            XSAV=ETA(IY)
          END IF
          DO 40 I=1,10
            DEL=S*DEL
            IF(I.EQ. 10.OR. ABS(XSAV+DEL*DX(9)-XSAV) .LT. EPS) THEN
              OK=.FALSE.
              RETURN
            END IF
            DO 10 K = 0,9
              H=DEL*DX(K)
              IF(LX) THEN
                X(IL)=XSAV+H
              ELSE
                ETA(IY)=XSAV+H
              END IF
              FPLUS=LSQGFN(ETA,X,N,NR,IM)
              IF(LX) THEN
                X(IL)=XSAV-H
              ELSE
                ETA(IY)=XSAV-H
              END IF
              FMINUS=LSQGFN(ETA,X,N,NR,IM)
              IF(LX) THEN
                X(IL)=XSAV
              ELSE
                ETA(IY)=XSAV
              END IF
              T(K,0)=(FPLUS-FMINUS)/(H+H)
              A(K)=T(K,0)
   10       CONTINUE
            IF(A(0) .GE. A(9)) THEN
              DO 20 K = 0,9
                A(K)=-A(K)
   20         CONTINUE
            END IF
            LMT=.TRUE.
            DO 30 K = 1,9
              H=A(K-1)-A(K)
              LMT=LMT .AND. (H.LE.EPSI .OR. ABS(H).LE.EPS*ABS(A(K))
     +        +EPSI)
   30       CONTINUE
            IF(LMT) GO TO 50
   40     CONTINUE
   50     CONTINUE
          DO 70 M = 1,9
            DO 60 K = 0,9-M
              IF(LEV(M)) THEN
                T(K,M)=W(M-1,1)*T(K+1,M-1)-W(M,1)*T(K,M-1)
              ELSE IF(LEV(K)) THEN
                T(K,M)=W(M-1,2)*T(K+1,M-1)-W(M,2)*T(K,M-1)
              ELSE
                T(K,M)=W(M-1,3)*T(K+1,M-1)-W(M,3)*T(K,M-1)
              END IF
   60       CONTINUE
   70     CONTINUE
          E(IM,I2)=T(0,9)
   80   CONTINUE
   90 CONTINUE
      END
